### PR TITLE
update an outdated error message in DDPPlugin

### LINKED ISSUE #8998: launching ddp fails with assertion error
## 🐛 Bug

Originally reported here: https://github.com/PyTorchLightning/pytorch-lightning/issues/4612#issuecomment-901539844

same for me, but the code doesn't get stuck at commit 03bb389 so it must be a change introduced afterwards

@williamFalcon (idk who else to tag) I tried to track down the problem by comparing the old commit and master.

```
plugins/training_type/ddp.py:131

     def setup_environment(self) -> None:
         # start the other scripts
-        if not self.cluster_environment.creates_children():
+        if not self.cluster_environment.creates_children() and os.environ.get("PL_IN_DDP_SUBPROCESS", "0") != "1":
             self._call_children_scripts()
```
removing os.environ... makes all created subprocesses go into _call_children_scripts() and then fail with assert local_rank == 0 causing the block, a solution would be reinstating the os check or adding:

```
if not self.cluster_environment.creates_children() and self.local_rank == 0:
    self._call_children_scripts()
```

### To Reproduce

TODO

### Environment

- PyTorch Lightning Version (e.g., 1.3.0):
- PyTorch Version (e.g., 1.8)
- Python version:
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #8998

Replaces an outdated error message saying that fit() is not allowed to be called multiple time. 
The new error message addresses a assertion error with a friendly message with possible causes of the malfunction.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -260,10 +259,11 @@ def setup_distributed(self):
         self.dist.device = self.root_device
 
     def _check_can_spawn_children(self):
-        if self._has_spawned_children:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp.py:263]: This was the only use of `_has_spawned_children`, remove it?

```python
$ grep -iIrn "_has_spawned_children" pytorch_lightning
pytorch_lightning/plugins/training_type/ddp.py:109:        self._has_spawned_children = False
pytorch_lightning/plugins/training_type/ddp.py:179:        self._has_spawned_children = True
pytorch_lightning/plugins/training_type/ddp.py:263:        if self._has_spawned_children:
```

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -260,10 +257,11 @@ def setup_distributed(self):
...
+        if self.local_rank != 0:
             raise RuntimeError(
-                "You tried to run `.fit` or `.test` multiple times in the same script."
-                " This is not supported in DDP mode, switch to `distributed_backend='ddp_spawn'` instead."
+                "Lightning attempted to launch new distributed processes from `local_rank > 0`. This should not happen."
```
Comment (kaushikb11) [inline pytorch_lightning/plugins/training_type/ddp.py:262]:
```suggestion
                "Lightning attempted to launch new distributed processes with `local_rank > 0`. This should not happen."
```

File: tests/plugins/test_ddp_plugin.py [hunk 3]
```diff
@@ -69,3 +72,25 @@ def test_ddp_barrier_non_consecutive_device_ids(barrier_mock, tmpdir):
...
+
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        accelerator="ddp",
+        num_processes=2,
+        plugins=[DDPPlugin(), WronglyImplementedEnvironment()],
+    )
+    with pytest.raises(
+        RuntimeError, match="Lightning attempted to launch new distributed processes from `local_rank > 0`."
```
Comment (awaelchli) [inline tests/plugins/test_ddp_plugin.py:94]:
```suggestion
        RuntimeError, match="Lightning attempted to launch new distributed processes with `local_rank > 0`."
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -57,6 +57,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added DeepSpeed Stage 1 support ([#8974](https://github.com/PyTorchLightning/pytorch-lightning/pull/8974))
 
 
+- Added a friendly error message when DDP attempts to spawn new distributed processes from rank > 0 ([#9005](https://github.com/PyTorchLightning/pytorch-lightning/pull/9005))
```
Comment (kaushikb11) [inline CHANGELOG.md:60]:
```suggestion
- Added a friendly error message when DDP attempts to spawn new distributed processes with rank > 0 ([#9005](https://github.com/PyTorchLightning/pytorch-lightning/pull/9005))
```

**Review by tchaton — APPROVED** [review]
LGTM !
